### PR TITLE
Apply event sourcing to Timer.CANCEL and Timer.TRIGGER processors

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR migrates the `Timer.CANCEL`, `Timer.CANCELLED`, `Timer.TRIGGER`, and `Timer.TRIGGERED` processors. The bulk of the changes is in the `TriggerTimerProcessor`, where most of the logic from `EventHandle` was pulled in (with some going into the applier). Another notable change is that `Timer.TRIGGERED` __must__ be applied _before_ `EVENT_OCCURRED`, as the applier will create the event trigger in the state which the event occurred processor needs to read. 

## Related issues

related to #6177 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by npepinpe** [comment]
Missing/left to do:

- [ ] ~Extend the random workflow generation (to do as a follow up)~
- [ ]  Add example test to assert that the state built during processing is the same as the one built during reprocessing (i.e. example property test)
- [ ] Discuss reuse of `EventHandle` after message subscription refactoring - there's definitely some duplicate code between the `TimerTrigger` and the way process instance correlation is handled.

**Comment by npepinpe** [comment]
Other notes:

I'm not super happy with the hack we did to pass around the side effects. Sometimes we have the interface `SideEffects`, sometimes it's a `Consumer<SideEffectProducer>`...I'd rather this was standardized.

The `TriggerTimerProcessor` definitely shares code with `EventHandle`, but `EventHandle` does not handle start events, I think (just the event sub processes, but not the process start events), so it wouldn't work as is. But I'm happy to discuss what we can do there.

File: engine/src/main/java/io/zeebe/engine/processing/timer/TriggerTimerProcessor.java [hunk 1]
```diff
@@ -70,72 +83,93 @@ public void processRecord(
...
+          .setBpmnElementType(BpmnElementType.START_EVENT)
+          .setFlowScopeKey(elementInstance.getKey());
+    } else {
+      eventOccurredKey = elementInstanceKey;
+    }
 
-      if (elementInstance != null && elementInstance.isActive()) {
-        return eventHandle.triggerEvent(streamWriter, elementInstance, catchEvent, NO_VARIABLES);
+    return eventOccurredKey;
+  }
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/processing/timer/TriggerTimerProcessor.java:143]: Following DRY, we should use `EventHandle.activateElement()` instead.

File: engine/src/main/java/io/zeebe/engine/processing/timer/TriggerTimerProcessor.java [hunk 2]
```diff
@@ -70,72 +83,93 @@ public void processRecord(
...
+    final long eventOccurredKey;
+    final var processInstanceKey = keyGenerator.nextKey();
+    eventOccurredKey = keyGenerator.nextKey();
+    eventOccurredRecord
+        .setBpmnElementType(BpmnElementType.START_EVENT)
+        .setProcessDefinitionKey(processDefinitionKey)
+        .setProcessInstanceKey(processInstanceKey)
+        .setElementId(timer.getTargetElementIdBuffer());
+    return eventOccurredKey;
+  }
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/processing/timer/TriggerTimerProcessor.java:155]: Following DRY, we should use `EventHandle.activateStartEvent()` instead.

File: engine/src/main/java/io/zeebe/engine/processing/timer/TriggerTimerProcessor.java [hunk 3]
```diff
@@ -70,72 +83,93 @@ public void processRecord(
...
-      final var processInstanceKey =
-          eventHandle.triggerStartEvent(
-              streamWriter, processDefinitionKey, timer.getTargetElementIdBuffer(), NO_VARIABLES);
-
-      return processInstanceKey > 0;
+      final var elementInstance = elementInstanceState.getInstance(elementInstanceKey);
+      result =
+          (elementInstance != null
+              && elementInstance.isActive()
+              && eventScopeInstanceState.isAcceptingEvent(elementInstanceKey));
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/processing/timer/TriggerTimerProcessor.java:166]: Following DRY, we should use `EventHandle.canTriggerElement()` instead.

File: engine/src/main/java/io/zeebe/engine/state/appliers/TimerTriggeredApplier.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (isStartEvent(elementInstanceKey)) {
+      eventScopeInstanceState.triggerEvent(
+          processDefinitionKey, key, targetElementId, NO_VARIABLES);
+    } else {
+      final ExecutableCatchEvent catchEvent =
+          processState.getFlowElement(
+              processDefinitionKey, targetElementId, ExecutableCatchEvent.class);
+      eventScopeInstanceState.triggerEvent(
+          elementInstanceKey, key, catchEvent.getId(), NO_VARIABLES);
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/state/appliers/TimerTriggeredApplier.java:54]: We don't need to look up for the element. The record value contains the element id. 

```suggestion
      eventScopeInstanceState.triggerEvent(
          elementInstanceKey, key, targetElementId, NO_VARIABLES);
```

**Review by saig0 — CHANGES_REQUESTED** [review]
Looks good :+1: 

I only have some suggestions to reduce duplication. Please have a look.

File: engine/src/main/java/io/zeebe/engine/state/appliers/TimerTriggeredApplier.java (hunk 4 above)
Comment (npepinpe) [reply engine/src/main/java/io/zeebe/engine/state/appliers/TimerTriggeredApplier.java:54]: Haha it seems so obvious now :see_no_evil:

**Comment by npepinpe** [comment]
I applied your review hints - most of the changes are in `TriggerTimerProcessor` :+1:

**Review by saig0 — APPROVED** [review]
:tada:

**Pull request merged**